### PR TITLE
Use CacheProvider instead of just implementing Cache

### PR DESCRIPTION
**Pull request by deviantintegral**
I'd like to be able to use the <a href="https://github.com/doctrine/cache/blob/master/lib/Doctrine/Common/Cache/ChainCache.php">Doctrine ChainCache</a> to stack cache providers. With ChainCache, you can easily create a static cache with an array cache, or a file cache, or anything else.

DrupalDoctrineCache only implements the Cache interface, but ChainCache requires an array of CacheProviders. I'm not sure the whole abstract class / method setup instead of an extended interface is best, but that's what we have to work with. Since CacheProvider implements Cache, this should be compatible with existing code.

The attached PR:
- Converts the class to use CacheProvider
- Switches to PSR-4
- Adds complete unit tests!
  - These intentionally just log function calls, since to be useful we'd really want _functional_ tests against a Drupal site. Replacing the calls with an ArrayCache just seemed even worse, given that we really just want to validate the functions are called correctly.

The diff is a bit large because of the PSR-4 conversion, but hopefully the individual commits make it easier to review. Once this is merged, the repo should be ready to configure with Travis / Coveralls etc for automated testing on PRs.

### CODE REVIEW

**Comment by tsphethean** [comment]
Thanks @deviantintegral - this is great stuff and makes sense. Will run some tests locally, but then should be good to merge.

File: src/DrupalCacheAdapter.php [hunk 1]
```diff
@@ -0,0 +1,23 @@
...
+namespace Capgemini\Cache;
+
+/**
+ * @class DrupalCacheAdapter
+ * @package Capgemini\Cache
+ * @codeCoverageIgnore
+ */
+trait DrupalCacheAdapter {
+
+  public function cache_get($cid, $bin = 'cache') {
```
Comment (alex-moreno) [inline src/DrupalCacheAdapter.php:41]: worth some docblocks in this class in general?

File: src/DrupalDoctrineCache.php [hunk 2]
```diff
@@ -26,82 +54,63 @@ public function fetch($id) {
   }
 
   /**
-   * Checks if the Drupal cache contains the specified entry.
```
Comment (alex-moreno) [inline src/DrupalDoctrineCache.php:29]: probably does not hurt if we keep the comments even if they have inherited?

File: src/DrupalDoctrineCache.php [hunk 3]
```diff
@@ -26,82 +54,63 @@ public function fetch($id) {
...
-   * Saves data to the Drupal cache.
-   *
-   * @inheritdoc
+   * {@inheritdoc}
    */
-  public function save($id, $data, $lifeTime = NULL) {
+  protected function doSave($id, $data, $lifeTime = 0) {
     // Doctrine defines NULL to mean permanent cache.
-    if ($lifeTime == NULL) {
+    if ($lifeTime === NULL) {
```
Comment (alex-moreno) [inline src/DrupalDoctrineCache.php:71]: if it is set to 0 by default, shouldn't we consistent and compare against 0 instead?

File: src/DrupalDoctrineCache.php [hunk 4]
```diff
@@ -26,82 +54,63 @@ public function fetch($id) {
...
-    $this->cache_table = $cache_table;
-  }
-
-  /**
-   * Get the cache table used by the caching API.
+   * Original description :-
    *
-   * @return string
+   * {@inheritdoc}
+   * @codeCoverageIgnore
```
Comment (alex-moreno) [inline src/DrupalDoctrineCache.php:116]: is that to avoid phpcs? Why in that case?

File: src/DrupalDoctrineCache.php [hunk 5]
```diff
@@ -26,82 +54,63 @@ public function fetch($id) {
...
-   */
-  public function setCacheTable($cache_table) {
-    $this->cache_table = $cache_table;
-  }
-
-  /**
-   * Get the cache table used by the caching API.
+   * Original description :-
    *
-   * @return string
```
Comment (alex-moreno) [inline src/DrupalDoctrineCache.php:101]: @return NULL?

File: tests/DrupalCacheAdapter.php [hunk 6]
```diff
@@ -0,0 +1,100 @@
...
+
+    return FALSE;
+  }
+
+  public function cache_set($cid, $data, $bin = 'cache', $expire = CACHE_PERMANENT) {
+    $args = func_get_args();
+    $this->logMethodCall(__FUNCTION__, $args);
+  }
+
+  public function cache_clear_all($cid = NULL, $bin = NULL, $wildcard = FALSE) {
```
Comment (alex-moreno) [inline tests/DrupalCacheAdapter.php:113]: docblocks?

File: tests/DrupalCacheAdapter.php [hunk 7]
```diff
@@ -0,0 +1,100 @@
...
+
+  /**
+   * Get the call count for a given key.
+   *
+   * @param string $key
+   *
+   * @return int
+   *   The number of times $key has been called.
+   */
+  public function getCallCount($key, $wildcard = FALSE) {
```
Comment (alex-moreno) [inline tests/DrupalCacheAdapter.php:128]: docblock is wrong (lacks wildcard and comments).

File: tests/DrupalCacheAdapter.php [hunk 8]
```diff
@@ -0,0 +1,100 @@
...
+    }
+
+    return 0;
+  }
+
+  /**
+   * @param $key
+   * @param $args
+   */
+  protected function getKey($key, $args) {
```
Comment (alex-moreno) [inline tests/DrupalCacheAdapter.php:157]: same here with docblocks, sorry.

File: tests/DrupalDoctrineCacheTest.php [hunk 9]
```diff
@@ -0,0 +1,65 @@
...
+   * @covers Capgemini\Cache\DrupalDoctrineCache::getCacheId
+   */
+  public function testCacheSet() {
+    $cache = new DrupalDoctrineCacheStub();
+    $cache->save('key', 'value');
+    $this->assertEquals(1, $cache->getCallCount('cache_set:doctrine:[key][1],value,cache,*', TRUE));
+    $cache->save('key', 'value', NULL);
+    $this->assertEquals(2, $cache->getCallCount('cache_set:doctrine:[key][1],value,cache,0'));
+    $cache->save('key', 'value', 60);
+    $this->assertEquals(1, $cache->getCallCount('cache_set:doctrine:[key][1],value,cache,' . substr(time(), 0, 1) . '*', TRUE));
```
Comment (alex-moreno) [inline tests/DrupalDoctrineCacheTest.php:34]: feels weird having to repeat that block. Worth creating a @dataProvider?

File: phpunit.xml.dist [hunk 10]
```diff
@@ -0,0 +1,17 @@
...
+    <testsuites>
+        <testsuite name="Drupal Doctrine Cache Test Suite">
+            <directory>./tests</directory>
+        </testsuite>
+    </testsuites>
+
+    <filter>
+        <blacklist>
+            <directory>./vendor</directory>
+            <directory>./tests</directory>
```
Comment (alex-moreno) [inline phpunit.xml.dist:13]: ./tests folder is added in both, testsuite and blacklist. Is that right?

File: src/DrupalCacheAdapter.php [hunk 11]
```diff
@@ -0,0 +1,23 @@
+<?php
+
+namespace Capgemini\Cache;
+
+/**
+ * @class DrupalCacheAdapter
+ * @package Capgemini\Cache
+ * @codeCoverageIgnore
```
Comment (alex-moreno) [inline src/DrupalCacheAdapter.php:18]: why do we want to ignore this?

File: src/DrupalCacheAdapter.php [hunk 12]
```diff
@@ -0,0 +1,23 @@
+<?php
+
```
Comment (alex-moreno) [inline src/DrupalCacheAdapter.php:2]: needs a @file description

File: src/DrupalCacheAdapter.php (hunk 11 above)
Comment (deviantintegral) [inline src/DrupalCacheAdapter.php:18]: This trait explicitly calls core's cache functions, which will require a bootstrapped Drupal instance.

File: src/DrupalDoctrineCache.php (hunk 4 above)
Comment (deviantintegral) [inline src/DrupalDoctrineCache.php:116]: We don't implement doGetStats(), so it didn't make sense to test it to me. The function could also remove the return to get the same effect?

File: phpunit.xml.dist (hunk 10 above)
Comment (deviantintegral) [inline phpunit.xml.dist:13]: Yes, this tells PHPUnit to ignore code coverage from the tests directory.

File: src/DrupalDoctrineCache.php (hunk 2 above)
Comment (deviantintegral) [inline src/DrupalDoctrineCache.php:29]: Adding back.

File: src/DrupalDoctrineCache.php (hunk 3 above)
Comment (deviantintegral) [inline src/DrupalDoctrineCache.php:71]: I originally did that, but then it turns out that if [you explicitly pass NULL in as $lifetime](https://php.net/manual/en/functions.arguments.php#70511) that the default argument doesn't get set. That's what [this test covers](https://github.com/Capgemini/drupal-doctrine-cache/pull/2/files#diff-a38057b53968471fe9b75cb89e824562R31).

File: tests/DrupalDoctrineCacheTest.php (hunk 9 above)
Comment (deviantintegral) [inline tests/DrupalDoctrineCacheTest.php:34]: I looked into this, but I'm not sure it makes the test any simpler. We have different save() parameters and different assertions in each case.

``` diff
diff --git a/tests/DrupalDoctrineCacheTest.php b/tests/DrupalDoctrineCacheTest.php
index 7ca9dbb..cee800f 100644
--- a/tests/DrupalDoctrineCacheTest.php
+++ b/tests/DrupalDoctrineCacheTest.php
@@ -20,18 +20,35 @@ class DrupalDoctrineCacheTest extends \PHPUnit_Framework_TestCase {
     $this->assertEquals('cache_test', $cache->getCacheTable());
   }

+  public function cacheSetProvider() {
+    return array(
+      array('key', 'value'),
+      array('key', 'value', NULL),
+      array('key', 'value', 60),
+    );
+  }
+
   /**
+   * @dataProvider cacheSetProvider
    * @covers Capgemini\Cache\DrupalDoctrineCache::doSave
    * @covers Capgemini\Cache\DrupalDoctrineCache::getCacheId
    */
-  public function testCacheSet() {
+  public function testCacheSet($key, $value, $expiry = 'unset') {
     $cache = new DrupalDoctrineCacheStub();
-    $cache->save('key', 'value');
-    $this->assertEquals(1, $cache->getCallCount('cache_set:doctrine:[key][1],value,cache,*', TRUE));
-    $cache->save('key', 'value', NULL);
-    $this->assertEquals(2, $cache->getCallCount('cache_set:doctrine:[key][1],value,cache,0'));
-    $cache->save('key', 'value', 60);
-    $this->assertEquals(1, $cache->getCallCount('cache_set:doctrine:[key][1],value,cache,' . substr(time(), 0, 1) . '*', TRUE));
+    if ($expiry == 'unset') {
+      $cache->save($key, $value);
+      $expiry = NULL;
+    }
+    else if ($expiry || $expiry == NULL) {
+      $cache->save($key, $value, $expiry);
+    }
+
+    if ($expiry) {
+      $this->assertEquals(1, $cache->getCallCount('cache_set:doctrine:[key][1],value,cache,' . substr(time(), 0, 1) . '*', TRUE));
+    }
+    else {
+      $this->assertEquals(1, $cache->getCallCount('cache_set:doctrine:[key][1],value,cache,0'));
+    }
   }

   /**
```

File: tests/DrupalCacheAdapter.php (hunk 7 above)
Comment (deviantintegral) [inline tests/DrupalCacheAdapter.php:128]: This is fixed, but github isn't updating the diff for me.

File: src/DrupalCacheAdapter.php (hunk 12 above)
Comment (deviantintegral) [inline src/DrupalCacheAdapter.php:2]: Fixed.

**Comment by deviantintegral** [comment]
I really should have run inspections before pushing this up. Sorry about that.

Anyways, I think everything is ready for review.

**Comment by alex-moreno** [comment]
As @tsphethean said, thank you very much for your contribution @deviantintegral :-)

**Comment by iainp999** [comment]
Why not just use delegation instead of replacing the existing Cache class?  That way, existing code will be backwards compatible too.

**Comment by iainp999** [comment]
Edit: misread the PR, sorry :)

**Comment by deviantintegral** [comment]
Let me know if there's any more followups on this. Once this is merged (and we have a tag?) I'd like to write up an article showing how to use this to replace manual drupal_static() / cache_set() calls.

**Comment by tsphethean** [comment]
It looks good @deviantintegral - planning on running a couple of tests with @alex-moreno today and then should be good to merge and tag.

The write up sounds great - would you be up for us cross-posting it on the Capgemini Engineering blog (capgemini.github.io) once you've published?

**Comment by deviantintegral** [comment]
@tsphethean that'd be great! We'd just want to be sure we don't get duplicate articles on the Drupal Planet from lullabot.com and capgemini.

**Comment by alex-moreno** [comment]
looking good, all tested in one of our envs and working like a charm :+1: 

Good to go if you want @tsphethean 

Thanks again @deviantintegral :-)

**Comment by tsphethean** [comment]
Great stuff, all merged. Thanks @deviantintegral for the work and @alex-moreno for the testing and review. I'll go prepare a new tag for packagist now.

@deviantintegral - happy for you to publish the post on lullabot.com for Drupal Planet, and then when we republish we'll make sure it's tagged so-as not to show up on Planet.

**Comment by deviantintegral** [comment]
I've got a draft of an article done. If anyone wants to review it, feel free to leave suggestions and comments. Thanks!

https://docs.google.com/document/d/1JzasLe66ps_eDa-nskDFPw-f4Y_q9VCkdSyFOloy-_g/edit?usp=sharing

**Comment by alex-moreno** [comment]
Being non English native, I can't say much about the style, apart of "good
job". Very nice, instructive article.

Regarding your last part, you may want to link to some libraries fully
integrated into D7 through composer, such as the Symfony inject one, which
brings all the magic of the Symfony container into drupal :
https://www.drupal.org/project/drupal_symfony_inject

Also related, that article that I wrote some weeks ago explains how we use
the container and composer :
http://capgemini.github.io/drupal/drupal-composer-symfony/

It's just a suggestion, please feel free to ignore me if it sounds too much
like self promoting (which probably it is :-( ).

Again, very good article, congrats Andrew and thanks a lot for sharing.

**Comment by iainp999** [comment]
Very nice Andrew, nothing to add from me.

**Comment by tsphethean** [comment]
Yes, I agree - great blog post and a really good idea.

**Comment by deviantintegral** [comment]
Ooo... I've never seen Symfony Inject before. I'll have to investigate that.

Thanks for the review everyone.

**Pull request merged**